airbyte-integrations/connectors/source-instagram/source_instagram/client/api.py

```diff
@@ -0,0 +1,268 @@
+        stream_name = self.__class__.__name__
+        if stream_name.endswith("API"):
+            stream_name = stream_name[:-3]
+        logger.info(f"Advancing bookmark for {stream_name} stream from {self._state} to {cursor}")
+        self._state = max(cursor, self._state) if self._state else cursor
+        return record
+
+
+class UsersAPI(StreamAPI):
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        yield self._extend_record(self._api.account, fields=fields)
+
+    def _extend_record(self, ig_user, fields) -> Dict:
+        return ig_user.api_get(fields=fields).export_all_data()
+
+
+class UserLifetimeInsightsAPI(StreamAPI):
+    LIFETIME_METRICS = ["audience_city", "audience_country", "audience_gender_age", "audience_locale"]
+    period = "lifetime"
+
+    buffer_days = 29
+    days_increment = 1
+
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        account_id = self._api.account.get("id")
+        for insight in self._get_insight_records(params=self._params()):
+            yield {
+                "id": account_id,
+                "metric": insight.get("name"),
+                "date": insight.get("values")[0]["end_time"],
+                "value": insight.get("values")[0]["value"],
+            }
+
+    def _params(self) -> Dict:
+        return {"metric": self.LIFETIME_METRICS, "period": self.period}
+
+    def _get_insight_records(self, params) -> Iterator[Any]:
+        return self._api.account.get_insights(params=params)
+
+
+class UserInsightsAPI(IncrementalStreamAPI):
+    METRICS_BY_PERIOD = {
+        "day": [
+            "email_contacts",
+            "follower_count",
+            "get_directions_clicks",
+            "impressions",
+            "phone_call_clicks",
+            "profile_views",
+            "reach",
+            "text_message_clicks",
+            "website_clicks",
+        ],
+        "week": ["impressions", "reach"],
+        "days_28": ["impressions", "reach"],
+        "lifetime": ["online_followers"],
+    }
+
+    state_pk = "date"
+
+    buffer_days = 29
+    days_increment = 1
+
+    def __init__(self, api):
+        super().__init__(api=api)
+        self._state = api._start_date
+
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        account_id = self._api.account.get("id")
+        for params_per_day in self._params():
+            insight_list = []
+            for params in params_per_day:
+                insight_list += self._get_insight_records(params=params)
+            if not insight_list:
+                continue
+
+            insight_record = {"id": account_id}
+            for insight in insight_list:
+                key = (
+                    f"{insight.get('name')}_{insight.get('period')}"
+                    if insight.get("period") in ["week", "days_28"]
+                    else insight.get("name")
+                )
+                insight_record[key] = insight.get("values")[0]["value"]
```

Comment:
is there always one value in the array?

Reply:
Yes, because now we are making a request to get data only 1 day in advance.

Reply:
Wouldn't it be more robust to output all items from the array without assuming there is one value in case we change something in the future or the api contract changes? Could we do that? And in the other locations too